File: SiHyeon/week6/체육복.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(int n, vector<int> lost, vector<int> reserve) {
    int answer = n - lost.size();
    sort(lost.begin(), lost.end());
    sort(reserve.begin(), reserve.end());
    vector<int> newLost;
    //여분 체육복을 도난 당한경우 미리 여분 체육복에서 제거
    for (int i = 0; i < lost.size(); i++)
    {
        auto index = find(reserve.begin(), reserve.end(), lost[i]);
        if (index != reserve.end())
        {
            reserve.erase(index);
            answer++;
        }
        else
            newLost.push_back(lost[i]); //도난당한 리스트를 갱신
    }
    //도난 당한 사람 앞뒤로 빌려줄 수 있는지 확인
    for (auto l : newLost)
    {
        auto index = find(reserve.begin(), reserve.end(), l - 1);
        if (index != reserve.end())
        {
            reserve.erase(index);
            answer++;
        }
        else
        {
            index = find(reserve.begin(), reserve.end(), l + 1);
            if (index != reserve.end())
            {
                reserve.erase(index);
                answer++;
            }

        }
    }
    return answer;
}

File: eunbi/week6/체육복.java
package algorithmStudy;
import java.util.*;

class Solution {
    public int solution(int n, int[] lost, int[] reserve) {
        List<Integer> l = new ArrayList<Integer>();
        List<Integer> r = new ArrayList<Integer>();
        
        for(int i : lost)
            l.add(i);
        for(int i : reserve)
            r.add(i);
        
        l.sort(Comparator.naturalOrder());
        r.sort(Comparator.naturalOrder());
        
        for(int i=0; i<r.size(); i++){
            if(l.contains(r.get(i))){ // 여벌 체육복이 있는데 도난 당한 경우
                    l.remove(l.indexOf(r.get(i)));
                    r.remove(i);
                    i--;
            }
        } // 이 경우를 먼저 계산해야함 (도난당한 학생은 빌려줄 수 없기 때문)
        
        for(int i=0; i<r.size(); i++){
            if(l.contains(r.get(i)-1)){ // 여분 학생의 앞 번호가 lost일 때
                l.remove(l.indexOf(r.get(i)-1));
                r.remove(i);
                i--; // list에서 제거했으므로 index를 이전으로 돌림
            }
            else if(l.contains(r.get(i)+1)){ // 여분 학생의 뒷 번호가 lost일 때
                l.remove(l.indexOf(r.get(i)+1));
                r.remove(i);
                i--;
            }
        }
        
        return n-l.size();
    }
}